File: packages/shared/src/billing/earnings.ts
import type { WorklogRow, ProjectEarning } from './types.js';
export type { ProjectEarning } from './types.js';

const inMonth = (workDate: string, month: string) => workDate.slice(0, 7) === month;
const isCzkEarned = (r: WorklogRow) => r.rateCurrency === 'CZK' && r.earnedAmount != null;

export function aggregateMonthEarnings(rows: WorklogRow[], month: string) {
  const byProject = new Map<number, ProjectEarning>();
  let totalCzk = 0;
  for (const r of rows) {
    if (!inMonth(r.workDate, month) || !isCzkEarned(r)) continue;
    totalCzk += r.earnedAmount!;
    const cur = byProject.get(r.projectId) ?? { projectId: r.projectId, name: r.projectName, color: r.projectColor, minutes: 0, earnedCzk: 0 };
    cur.minutes += r.minutes;
    cur.earnedCzk += r.earnedAmount!;
    byProject.set(r.projectId, cur);
  }
  const perProject = [...byProject.values()].sort((a, b) => b.earnedCzk - a.earnedCzk);
  return { totalCzk, perProject };
}

function addMonths(month: string, delta: number): string {
  const parts = month.split('-');
  const y = Number(parts[0]);
  const m = Number(parts[1]);
  const d = new Date(Date.UTC(y, m - 1 + delta, 1));
  return `${d.getUTCFullYear()}-${String(d.getUTCMonth() + 1).padStart(2, '0')}`;
}

export function trailingMonths(rows: WorklogRow[], endMonth: string, n: number) {
  const months = Array.from({ length: n }, (_, i) => addMonths(endMonth, -(n - 1 - i)));
  const totals = new Map(months.map((m) => [m, 0]));
  for (const r of rows) {
    if (!isCzkEarned(r)) continue;
    const m = r.workDate.slice(0, 7);
    if (totals.has(m)) totals.set(m, totals.get(m)! + r.earnedAmount!);
  }
  return months.map((month) => ({ month, earnedCzk: totals.get(month)! }));
}

export function topProjects(rows: WorklogRow[], month: string, limit: number) {
  const by = new Map<number, { projectId: number; name: string; color: string | null; minutes: number; earnedCzk: number }>();
  for (const r of rows) {
    if (!inMonth(r.workDate, month)) continue;
    const cur = by.get(r.projectId) ?? { projectId: r.projectId, name: r.projectName, color: r.projectColor, minutes: 0, earnedCzk: 0 };
    cur.minutes += r.minutes;
    if (isCzkEarned(r)) cur.earnedCzk += r.earnedAmount!;
    by.set(r.projectId, cur);
  }
  return [...by.values()]
    .filter((p) => p.minutes > 0)
    .sort((a, b) => b.minutes - a.minutes || a.name.localeCompare(b.name))
    .slice(0, limit);
}

File: swift/WatchtowerCore/Sources/WatchtowerCore/Billing/Earnings.swift
import Foundation

private func inMonth(_ workDate: String, _ month: String) -> Bool {
    String(workDate.prefix(7)) == month
}

private func isCzkEarned(_ r: WorklogRow) -> Bool {
    r.earnedAmount != nil
}

/// Port of `packages/shared/src/billing/earnings.ts` `aggregateMonthEarnings`.
/// `perProject` is sorted by `earnedCzk` descending; `totalCzk` sums only
/// rows with non-nil `earnedAmount` in `month`.
public func aggregateMonthEarnings(_ rows: [WorklogRow], _ month: String) -> (totalCzk: Double, perProject: [ProjectEarning]) {
    var order: [Int] = []
    var byProject: [Int: ProjectEarning] = [:]
    var totalCzk: Double = 0

    for r in rows {
        guard inMonth(r.workDate, month) else { continue }
        var cur = byProject[r.projectId] ?? ProjectEarning(projectId: r.projectId, name: r.projectName, color: r.projectColor, minutes: 0, earnedCzk: 0)
        if byProject[r.projectId] == nil {
            order.append(r.projectId)
        }
        cur = ProjectEarning(projectId: cur.projectId, name: cur.name, color: cur.color,
                              minutes: cur.minutes + r.minutes, earnedCzk: cur.earnedCzk)
        if isCzkEarned(r) {
            cur = ProjectEarning(projectId: cur.projectId, name: cur.name, color: cur.color,
                                  minutes: cur.minutes, earnedCzk: cur.earnedCzk + r.earnedAmount!)
            totalCzk += r.earnedAmount!
        }
        byProject[r.projectId] = cur
    }

    var position: [Int: Int] = [:]
    for (i, pid) in order.enumerated() { position[pid] = i }
    let perProject = order.compactMap { byProject[$0] }
        .sorted {
            $0.earnedCzk != $1.earnedCzk
                ? $0.earnedCzk > $1.earnedCzk
                : (position[$0.projectId] ?? 0) < (position[$1.projectId] ?? 0)
        }
    return (totalCzk, perProject)
}

/// Port of `trailingMonths`: the `n` months ending at `endMonth` (ascending),
/// each summing non-nil `earnedAmount`, zero-filled for months with no rows.
public func trailingMonths(_ rows: [WorklogRow], _ endMonth: String, _ n: Int) -> [(month: String, earnedCzk: Double)] {
    let months = (0..<n).map { CzFormat.addMonths(endMonth, -(n - 1 - $0)) }
    var totals: [String: Double] = Dictionary(uniqueKeysWithValues: months.map { ($0, 0) })

    for r in rows {
        guard isCzkEarned(r) else { continue }
        let m = String(r.workDate.prefix(7))
        if totals[m] != nil {
            totals[m]! += r.earnedAmount!
        }
    }

    return months.map { (month: $0, earnedCzk: totals[$0]!) }
}

/// Port of `topProjects`: filter `minutes > 0`, sort `minutes` desc then
/// `name` ascending, take `limit`.
public func topProjects(_ rows: [WorklogRow], _ month: String, _ limit: Int) -> [ProjectEarning] {
    var order: [Int] = []
    var byProject: [Int: ProjectEarning] = [:]

    for r in rows {
        guard inMonth(r.workDate, month) else { continue }
        var cur = byProject[r.projectId] ?? ProjectEarning(projectId: r.projectId, name: r.projectName, color: r.projectColor, minutes: 0, earnedCzk: 0)
        if byProject[r.projectId] == nil {
            order.append(r.projectId)
        }
        cur = ProjectEarning(projectId: cur.projectId, name: cur.name, color: cur.color,
                              minutes: cur.minutes + r.minutes, earnedCzk: cur.earnedCzk)
        if isCzkEarned(r) {
            cur = ProjectEarning(projectId: cur.projectId, name: cur.name, color: cur.color,
                                  minutes: cur.minutes, earnedCzk: cur.earnedCzk + r.earnedAmount!)
        }
        byProject[r.projectId] = cur
    }

    var position: [Int: Int] = [:]
    for (i, pid) in order.enumerated() { position[pid] = i }
    return order.compactMap { byProject[$0] }
        .filter { $0.minutes > 0 }
        .sorted {
            if $0.minutes != $1.minutes { return $0.minutes > $1.minutes }
            if $0.name != $1.name { return $0.name.localizedCompare($1.name) == .orderedAscending }
            return (position[$0.projectId] ?? 0) < (position[$1.projectId] ?? 0)
        }
        .prefix(limit)
        .map { $0 }
}
